File: Python/preorderTraversal.py
#!/usr/bin/env python3
# encoding: utf-8

"""
@Filename: preorderTraversal.py
@Function: 二叉树的前序遍历
@Link: https://leetcode-cn.com/problems/binary-tree-preorder-traversal/
@Python Version: 3.8
@Author: Wei Li
@Date: 2021-07-31
"""


from typing import List

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def preorderTraversal(self, root: TreeNode) -> List[int]:
        res = list()
        if not root:
            return res

        stack = []
        node = root
        while stack or node:
            while node:
                res.append(node.val)
                stack.append(node)
                node = node.left
            node = stack.pop()
            node = node.right
        return res


# ---------------------------------
if __name__ == "__main__":
    # test on LeetCode online.
    pass

File: C++/preorderTraversal.cpp
/**
 * @File    : preorderTraversal.cpp
 * @Brief   : 二叉树的前序遍历
 * @Link    : https://leetcode-cn.com/problems/binary-tree-preorder-traversal/
 * @Author  : Wei Li
 * @Date    : 2021-07-31
*/

#include <iostream>
#include <vector>
#include <stack>

// Definition for a binary tree node.
struct TreeNode
{
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution
{
public:
    std::vector<int> preorderTraversal(TreeNode *root)
    {
        std::vector<int> ret;
        if (!root)
        {
            return ret;
        }
        std::stack<TreeNode *> s;
        s.push(root);
        while (!s.empty())
        {
            TreeNode *node = s.top();
            s.pop();
            ret.push_back(node->val);
            if (node->right)
            {
                s.push(node->right); // 先右后左， 保证左子树先遍历
            }
            if (node->left)
            {
                s.push(node->left);
            }
        }
        return ret;
    }
};

// -------------------------------
int main(int argc, char **argv)
{
    // test on LeetCode online.

    return 0;
}
